cache/remotecache/azblob/exporter.go

```diff
@@ -0,0 +1,214 @@
+			config:              config,
+		}, nil
+	}
+}
+
+var _ remotecache.Exporter = &exporter{}
+
+type exporter struct {
+	solver.CacheExporterTarget
+	chains          *v1.CacheChains
+	containerClient *container.Client
+	config          *Config
+}
+
+func (ce *exporter) Name() string {
+	return "exporting cache to Azure Blob Storage"
+}
+
+func (ce *exporter) Finalize(ctx context.Context) (map[string]string, error) {
+	config, descs, err := ce.chains.Marshal(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	for i, l := range config.Layers {
+		dgstPair, ok := descs[l.Blob]
+		if !ok {
+			return nil, errors.Errorf("missing blob %s", l.Blob)
+		}
+		if dgstPair.Descriptor.Annotations == nil {
+			return nil, errors.Errorf("invalid descriptor without annotations")
+		}
+		var diffID digest.Digest
+		v, ok := dgstPair.Descriptor.Annotations[labels.LabelUncompressed]
+		if !ok {
+			return nil, errors.Errorf("invalid descriptor without uncompressed annotation")
+		}
+		dgst, err := digest.Parse(v)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to parse uncompressed annotation")
+		}
+		diffID = dgst
+
+		key := blobKey(ce.config, dgstPair.Descriptor.Digest.String())
+
+		exists, err := blobExists(ctx, ce.containerClient, key)
+		if err != nil {
+			return nil, err
+		}
+
+		bklog.G(ctx).Debugf("layers %s exists = %t", key, exists)
+
+		if !exists {
+			layerDone := progress.OneOff(ctx, fmt.Sprintf("writing layer %s", l.Blob))
+			ra, err := dgstPair.Provider.ReaderAt(ctx, dgstPair.Descriptor)
+			if err != nil {
+				err = errors.Wrapf(err, "failed to get reader for %s", dgstPair.Descriptor.Digest)
+				return nil, layerDone(err)
+			}
+			if err := ce.uploadBlobIfNotExists(ctx, key, content.NewReader(ra)); err != nil {
+				return nil, layerDone(err)
+			}
+			layerDone(nil)
+		}
+
+		la := &v1.LayerAnnotations{
+			DiffID:    diffID,
+			Size:      dgstPair.Descriptor.Size,
+			MediaType: dgstPair.Descriptor.MediaType,
+		}
+		if v, ok := dgstPair.Descriptor.Annotations["buildkit/createdat"]; ok {
+			var t time.Time
+			if err := (&t).UnmarshalText([]byte(v)); err != nil {
+				return nil, err
+			}
+			la.CreatedAt = t.UTC()
+		}
+		config.Layers[i].Annotations = la
+	}
+
+	dt, err := json.Marshal(config)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to marshal config")
+	}
+
+	for _, name := range ce.config.Names {
+		if innerError := ce.uploadManifest(ctx, manifestKey(ce.config, name), bytesToReadSeekCloser(dt)); innerError != nil {
+			return nil, errors.Wrapf(innerError, "error writing manifest %s", name)
+		}
+	}
+
+	return nil, nil
+}
+
+func (ce *exporter) Config() remotecache.Config {
+	return remotecache.Config{
+		Compression: compression.New(compression.Default),
+	}
+}
+
+// For uploading manifests, use the Upload API which follows "last writer wins" sematics
+// This is slightly slower than UploadStream call but is safe to call concurrently from multiple threads. Refer to:
```

Comment:
Do you need to check our `UploadStream` call in the gha backend?

Reply:
ok to do in a follow up PR? wanted to keep this limited to the vulnerability fixes so keeping everything else as it was. need to do some more testing (talking to azblob team) before changing this